### ISSUE TITLE
ExecutorBeanPostProcessor creates proxies with null fields values

### ISSUE THREAD

**Issue by novy1234** [spring-cloud/spring-cloud-sleuth#1336]
Hi,
I found a problem during creation of async executors proxies in ExecutorBeanPostProcessor. 
Namely when creating proxies for the executors they are created with null fields inside, which is a bit problematic if you use any delegating executor, e.g. DelegatingSecurityContextSchedulingTaskExecutor. 

From what I have seen, CGLib uses no-args constructor if no constructor arguments have been passed, hence the fields have null values

I use Greenwich.SR1 train/release

Can you help me with that?

**Comment by marcingrzejszczak** [comment]
All CGLIB proxies have null fields. You shouldn't be using those. What is exactly your problem?

**Comment by novy1234** [comment]
@marcingrzejszczak My problem is that ExecutorBeanPostProcessor creates proxy of DelegatingSecurityContextSchedulingTaskExecutor which during execution forwards call to yet another task executor. It cannot do so, because the delegate field is null

**Comment by marcingrzejszczak** [comment]
ok, can you create a small sample that replicates this issue? ALso can you check the latest snapshots, cause AFAIR I've added some fixes at some point

**Comment by roll57** [comment]
Hello,
I have the same issue, I notice the following log when I start my server:

> o.s.aop.framework.CglibAopProxy          : Unable to proxy interface-implementing method [public final java.util.List org.springframework.security.concurrent.DelegatingSecurityContextExecutorService.invokeAll(java.util.Collection,long,java.util.concurrent.TimeUnit) throws java.lang.InterruptedException] because it is marked as final: Consider using interface-based JDK proxies instead!

In my code I do:
`executorService.invokeAll(...)`

I tried with an executorService which have an _invokeAll_ method that is not final and it works.

**Comment by marcingrzejszczak** [comment]
We can't wrap final methods with cglib. Just try no to use them if possible. My request for a sample is still active

**Comment by kilink** [comment]
[DelegatingSecurityContextAsyncTaskExecutor](https://github.com/spring-projects/spring-security/blob/master/core/src/main/java/org/springframework/security/task/DelegatingSecurityContextAsyncTaskExecutor.java) has such final methods; registering one as a bean and and then trying to use it anywhere in my app, I get the same issue.

```java
@Configuration
public class AppConfig {
    @Bean
    public Executor executor(TaskExecutorBuilder: builder) {
        return new DelegatingAsyncTaskExecutor(builder.build());
    }
}
```

This is something that worked fine before 2.1.1. I need to use the above decorator, so I guess my only option is to not use spring-cloud-sleuth?

**Comment by roll57** [comment]
Your other option is to do your own decorator.
```
public class MyExecutorServiceDecorator implements ExecutorService {

    private final ExecutorService delegate;

    public OcpExecutorServiceDecorator(ExecutorService executorService) {
        this.delegate = executorService;
    }

    @Override
    public void shutdown() {
        delegate.shutdown();
    }

    @Override
    public List<Runnable> shutdownNow() {
        return delegate.shutdownNow();
    }

    @Override
    public boolean isShutdown() {
        return delegate.isShutdown();
    }

    @Override
    public boolean isTerminated() {
        return delegate.isTerminated();
    }

    @Override
    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
        return delegate.awaitTermination(timeout, unit);
    }

    @Override
    public <T> Future<T> submit(Callable<T> task) {
        return delegate.submit(task);
    }
....
```

**Comment by marcingrzejszczak** [comment]
Can someone create a test that replicates this issue? I'd love to help but I need to have something to work on.

**Comment by roll57** [comment]
I replicate the problem in this repository 
https://github.com/roll57/scs1336

**Comment by kilink** [comment]
I've opened a PR with a proposed fix for this.

I noticed that the wrapExecutor explicitly checked if the "execute" method is final or not, in addition to checking if the class if final, when determining whether or not to use a cglib proxy. All of the other relevant methods did not have this check at all, and instead only checked if the class is final.

The proposed fix will check if the bean has any final methods, and use the fallback jdk proxy if there are. The one issue I encountered was with ThreadPoolTaskExecutor, since it is not an interface, proxying won't really work at all if there are any final methods. Therefore I changed the code just to always fallback to the Sleuth wrapper implementation.
